### PR TITLE
Multiuser Support

### PR DESCRIPTION
**Pull request by cantino**
@rishabhjain excellent work so far on multiuser support.

### CODE REVIEW

**Comment by adambedford** [comment]
This looks like a good start and associations are now fetched correctly. Seems like save operations still does not use the correct client object.

**Comment by cantino** [comment]
@adambedford Good to hear!  Can you commit any improvements to handle save?

**Comment by rishabhjain** [comment]
@adambedford I committed a fix. Can you check now.

**Comment by adambedford** [comment]
@rishabhjain That commit fixed the save to a certain extent but I'm noticing a couple of other items. 

`Mavenlink::Client.new(oauth_token: 'some_token').workspaces.create` returns a `Mavenlink::Response` object rather than a model object (`Mavenlink::Workspace` in this case). Also, related to that, `Mavenlink::Client.new(oauth_token: 'some_token').workspaces.new` tries to initialize a `Mavenlink::Response` object rather than a `Mavenlink::Workspace` object.

**Comment by rishabhjain** [comment]
@adambedford On `master` I'm getting the same issue. I think this bug has been present in `master` all along. `Mavenlink::Response` object has `client` attribute, so we can wrap it into `Mavenlink::Model`. I'll look into it.

File: lib/mavenlink/model.rb [hunk 1]
```diff
@@ -111,14 +118,21 @@ def self.update_attributes
...
+      self.new({}, source_record, client)
     end
 
     # @param attributes [Hash]
     # @param source_record [BrainstemAdaptor::Record]
-    def initialize(attributes = {}, source_record = nil)
+    # @param client [Mavenlink::Client]
+    def initialize(attributes = {}, source_record = nil, client = nil)
       super(self.class.collection_name, (attributes[:id] || attributes['id'] || source_record.try(:id)), source_record.try(:response))
+      if client
```
Comment (einzige) [inline lib/mavenlink/model.rb]: I believe this is `@client = client || Mavenlink.client`.

File: lib/mavenlink/model.rb [hunk 2]
```diff
@@ -36,6 +38,11 @@ def self.scoped
       Mavenlink::Request.new(collection_name)
     end
 
+    # @return [Mavenlink::Request]
+    def scoped_im
```
Comment (einzige) [inline lib/mavenlink/model.rb]: why public?

File: lib/mavenlink/model.rb (hunk 1 above)
Comment (einzige) [inline lib/mavenlink/model.rb]: i believe this should be dome this way:

```ruby
def init(attrs = {}, source = nil, client: Mavenlink.client)
    @client = client
end
```

can't find out why you use `||=`

File: spec/lib/mavenlink/model_spec.rb [hunk 3]
```diff
@@ -175,6 +175,11 @@ def self.name
     it 'accepts attributes' do
       expect(model.new(any_custom_key: 'value set')).to include(any_custom_key: 'value set')
     end
+
+    it 'checks for client' do
+      expect(model.new).to respond_to (:client)
```
Comment (einzige) [inline spec/lib/mavenlink/model_spec.rb]: no need for this one

File: spec/lib/mavenlink/model_spec.rb [hunk 4]
```diff
@@ -175,6 +175,11 @@ def self.name
...
+      expect(model.new.client).to be_a Mavenlink::Client
+    end
```
Comment (einzige) [inline spec/lib/mavenlink/model_spec.rb]: expect model.new(...client).client to eq (client)

File: spec/lib/mavenlink/response_spec.rb [hunk 5]
```diff
@@ -23,9 +23,14 @@
       expect(response.results).to have(2).records
     end
 
+    it 'checks for client' do
+      expect(response).to respond_to (:client)
```
Comment (einzige) [inline spec/lib/mavenlink/response_spec.rb]: l28  checks that

File: spec/lib/mavenlink/response_spec.rb [hunk 6]
```diff
@@ -23,9 +23,14 @@
...
+      expect(response.client).to be_a Mavenlink::Client
```
Comment (einzige) [inline spec/lib/mavenlink/response_spec.rb]: check for assignment through constructor (same as for model)

File: spec/lib/mavenlink/workspace_spec.rb [hunk 7]
```diff
@@ -51,6 +51,14 @@
     end
   end
 
+  describe 'instance methods' do
+    subject { model.new(title: 'Some title', creator_role: 'maven') }
+
+    specify do
+      expect(subject.scoped_im).to be_a Mavenlink::Request
```
Comment (einzige) [inline spec/lib/mavenlink/workspace_spec.rb]: this method serves internal interface, we don't need this to be in public

File: lib/mavenlink/model.rb (hunk 2 above)
Comment (einzige) [inline lib/mavenlink/model.rb]: rename to "request_scope" or "base_request"?

**Comment by cantino** [comment]
@adambedford Could you like to take ownership of this effort, since it's affecting you?  @rishabhjain was kind enough to donate his time as part of an interview, but that engagement isn't currently moving forward, so he won't be available for further changes.

**Comment by adambedford** [comment]
@cantino Sure. I'll open a new PR that addresses some this these comments.

**Comment by cantino** [comment]
@adambedford, if it would be helpful, @robyurkowski can take ownership of this.

**Comment by cantino** [comment]
Wait, has this effectively been fixed by https://github.com/mavenlink/mavenlink_gem/pull/21 ?

**Comment by einzige** [comment]
@cantino yep, we can close this one

**Pull request closed without merging**